### ISSUE TITLE
Performance Tips for Categorical Data

### ISSUE THREAD

**Issue by jonas-schulze** [bkamins/Julia-DataFrames-Tutorial#11]
In your [performance tips](
https://render.githubusercontent.com/view/ipynb?commit=32a08895be41fc20c9f8c9da408909aa404f806c&enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f626b616d696e732f4a756c69612d446174614672616d65732d5475746f7269616c2f333261303838393562653431666332306339663863396461343038393039616134303466383036632f31315f706572666f726d616e63652e6970796e62&nwo=bkamins%2FJulia-DataFrames-Tutorial&path=11_performance.ipynb&repository_id=112856111&repository_type=Repository#Allowing-missing-as-well-as-categorical-slows-down-computations) at `In [8]` you mentioned:
```
Allowing missing as well as categorical slows down computations
```
I might have an idea on how to fix this (at least partially). I did not do any research on whether there has already been done some work on this. If we compute the count map on the references into the pool, we only operate on `UInt`s:
```julia
julia> function StatsBase.countmap(x::CategoricalArray; alg = :auto)
           cm = countmap(x.refs, alg=alg)
           Dict(x.pool[ref] => count for (ref, count) in cm)
       end

julia> x = rand(1:10, 10^6);

julia> y = categorical(x);

julia> z = compress(y);

julia> @btime countmap(x);
  2.927 ms (8 allocations: 7.63 MiB)

julia> @btime countmap(y);
  2.141 ms (14 allocations: 3.82 MiB)

julia> @btime countmap(z);
  541.943 μs (11 allocations: 3.27 KiB)
```
On top of that, if we guard `x.pool[ref]` for `ref == 0`, which corresponds to a `missing` value, we also fix the performance loss for missing values.

My question is, where should I propose and contribute a change like the above, `CategoricalArrays.jl` or `StatsBase.jl` or somewhere else? Neither of the two has the other as a dependency, but they need to be added to allow for type dispatch (I think; I'm still a Julia novice), so maybe it should go to `DataFrames.jl`?

**Comment by bkamins** [comment]
`countmap` in this tutorial is just an example.

Regarding your proposal it should rather go to CategoricalArrays.jl as otherwise it would be type piracy (also because you use direct field access which should not be done outside CategoricalArrays.jl).

@nalimilan - do you think it is worth to add this specific method (maybe a review of other standard methods would be in place?).

**Comment by jonas-schulze** [comment]
> `countmap` in this tutorial is just an example.

I know, but from the top of my head, it should also be trivial to optimize `map`, `sum` and `prod`.

> it should rather go to CategoricalArrays.jl as otherwise it would be type piracy

Right. But how would one cope with the fact, that `CategoricalArrays.jl` otherwise does not rely on or need `StatsBase.jl` in any way. Every user of `CategoricalArrays.jl` would inherit that "blind" dependency.

If you give me some pointers on where to put these patches, I'd be happy to help. 🙂

**Comment by nalimilan** [comment]
FreqTables also has an optimized method, but I couldn't add it to StatsBase's `countmap` to avoid having it depend on CategoricalArrays. Maybe we could use `@require`, though. The same issue applies to PooledArrays.

**Comment by bkamins** [comment]
Having thought about it my conclusion is that it should be acceptable even to add StatsBase.jl as a dependency of CategoricalArrays.jl. The reason is that `CategoricalArray` is a statistics-oriented type.

This opposed to PooledArrays.jl which are general and should not depend on StatsBase.jl I think.

But in both cases we could consider using `@require`. My only question with this would be wouldn't then the behavior depend on code loading sequence?

**Comment by nalimilan** [comment]
> But in both cases we could consider using `@require`. My only question with this would be wouldn't then the behavior depend on code loading sequence?

Not AFAIK, that's the strength of Requires.jl.

One thing to keep in mind is that StatsBase is supposed to be merged into Statistics at some point, and then it won't be possible to use `@require` I guess. But at that point it won't be a problem for CategoricalArrays and PooledArrays to depend on Statistics.

**Comment by xiaodaigh** [comment]
Does the invention of DataAPI.jl change anything here.

**Comment by nalimilan** [comment]
Yes, it should be possible to add an optimized code path using `refarray`.
